Fix the client's available messages counter.

batiati opened the pull request:
A subtle bug when a request is timed-out and its reply arrives when it is being retransmitted, causes the counter not to be incremented because the request is still referenced by the send queue.

The most common effect could be observed during `client.deinit()` when the assertion `self.messages_available == constants.client_request_queue_max` failed.

- Reworks `client.messages_available` to not rely on the message's reference counter.

- Renames `client.unref` to `client.release` and changes the call site. When a message is acquired from `client.get_message`, it should never call `message.ref()` or `message_bus.unref()` directly.

- Adds `-Dconfig=production` in all the client's builds.

- Adds the missing `TooMuchData` test to the dotnet unit tests.

- Adds support to the environment variable `TB_ADDRESS` to the `c_sample`.
It allows the C sample to work along `run_with_tb`.

Closes #1206

batiati commented on src/vsr/client.zig:
```diff
@@ -293,11 +266,58 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         pub fn unref(self: *Self, message: *Message) void {
+            // Only request messages are unreferenced by this function,
+            // all others should use `message_bus.unref` directly.
+            assert(message.header.command == .request);
```
This assumption that `unref` can only accept `.request` heavily depends on how `tb_client` works.
One using `vsr.client` directly can call `get_message` and then `unref` without setting the message header, causing a panic for accessing undefined memory.

Even so, I believe this assertion is worth keeping to avoid misuse between `client.unref`, `bus.unref` and `pool.unref`.

batiati replied:
Assertion removed, since `client.release` can be called by `errdefer` with an uninitialized message.

batiati commented on src/vsr/client.zig:
```diff
@@ -151,37 +154,8 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-                    });
-                    return;
-                },
-            }
```
Assuming that it's possible for a client to receive a misdirected `.request`, it would incorrectly increment the message counter when unref`d.
We could mitigate it by checking the client and cluster id, but it's still possible to receive a request formatted like that (I can't think of a way for it to happen without a malicious agent, but it's not impossible).

I'm starting to wonder if it's worth keeping `client.messages_available`, or if we could just rely on `client.request_queue.count` to check if we are out of messages.

Assuming that we have calculated the correct maximum number of messages to never exhaust the message pool, even in odds cases such as retransmission (the message is still held by the send queue after leaving the request queue).

batiati replied:
> I can't think of a way for it to happen without a malicious agent, but it's not impossible

😂 Just discovered a particular case when the `recv_message` is unref'd during `deinit`, causing the message counter to be double incremented because the message header was left with the previous `command == .request`.

See the commit message for more info.

batiati committed:
Fix client messages_available counter.

- Reworks `client.messages_available` to not rely on the message's reference counter.

- Renames `client.unref` to `client.release` and changes the call site. When a message is acquired from `client.get_message`, it should never call `message.ref()` or `message_bus.unref()` directly.

- Adds `-Dconfig=production` in all the client's builds.

- Adds the missing `TooMuchData` test to the dotnet unit tests.

- Adds support to the environment variable `TB_ADDRESS` to the `c_sample`.
It allows the C sample to work along `run_with_tb`.

batiati committed:
Fix simulator

batiati committed:
Fix node client

batiati committed:
Fix demos

matklad commented on src/vsr/client.zig:
```diff
@@ -292,11 +292,10 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         }
 
         /// Releases a message back to the message bus.
-        pub fn unref(self: *Self, message: *Message) void {
```
Let's adjust the docs for get_message to say that the message must end up in release or request, and that it must _not_ be `ref`ed, as this flow is different from how one typically uses the `Message`. "`.ref()` if you store the message" is the general rule we use elsewhere, and here we are deliberately violating that.

matklad commented on src/vsr/client.zig:
```diff
@@ -388,12 +387,13 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
 
-            // Even though we release our reference to the message, the user might have retained
+            // Even though we release our reference to the message, we might have retained
             // another one.
```
Let's mention the `send` queue here; as written, the reader will wonder how _exactly_ do we retain the message.

matklad commented on src/vsr/client.zig:
```diff
@@ -234,7 +234,7 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
                 .user_data = user_data,
                 .callback = callback,
-                .message = message.ref(),
+                .message = message,
```
I am debating whether we should add `errdefer self.release` for readability and for simmetry with register. I think we'd better not.

batiati replied:
I prefer the pattern `errdefer self.release()` only after `client.get_message()`.

matklad commented on src/clients/java/pom.xml:
```diff
@@ -243,6 +243,7 @@
                 <argument>build</argument>
                 <argument>java_client</argument>
                 <argument>-Doptimize=ReleaseSafe</argument>
+                <argument>-Dconfig=production</argument>
```
Such a great find!

batiati replied:
It was with @kassane 's help!
Thanks for reporting the bug.

batiati committed:
- Better comments.

- Fixes the benchmark to not store the message reference.

- Uses const where applicable.
